### DISCUSSION TITLE
How to handle design when parameterization

### DISCUSSION THREAD

**Discussion by skrabbe** [QuestPDF/QuestPDF#480] [Q&A] [answered]
We are using QuestPDF for creating invoices (AWESOME library! :)).
Since we are hosting our application for several customers the layout of the invoices needs to be parameterization due to customer demands.
I have a question about how to handle design changes, we are returning from database invoice structure (rows, tables, columns..) and invoice design, questpdf then creates elements and sets design according to this.
For example;

Loop elements and create new cell for each child, use static CreateTableStyle, returns ITableCellContainer
` foreach (var child in database)
                            {
                                table.CreateTableStyle(child.Design).Component(new TTypeComponent(child, Documents));
                            }`

CreateTableStyle take design interface
`public static IContainer CreateTableStyle(this TableDescriptor item, TableDesign design)`

One of the properties are how to align in cell
` var align = design?.Align;
            var aCenter = align?.Center ?? false;
            var aMiddle = align?.Middle ?? false;
            var aRight = align?.Right ?? false;`

If cell needs to be aligned center, since container does not have container.Align(x => aCenter ? "center" : x) I solved it with AlignCenter and set container from return
` if (aCenter)
                container = container.AlignCenter();`
Is this correct? Could I have solved it differently? I get alot of this code due to design-checks... if-cases to check for align center/left/right/top/bottom and more...

Last code of CreateTableStyle, return IContainer from CreateBasicElement stru
`return container.CreateBasicElement(design);`

Static method CreateBasicElement
`public static IContainer CreateBasicElement<TParent>(this TParent parent, TableDesign design) where TParent`

For example, get any padding
` var paddingTop = design?.Padding?.Top ?? 0;`

Return element and set design properties according to interface
`  return parent
                .Element(x => paddingTop != 0 ? x.PaddingTop(paddingTop) : x)
                .Element(x => paddingBottom != 0 ? x.PaddingBottom(paddingBottom) : x)
                .Element(x => paddingLeft != 0 ? x.PaddingLeft(paddingLeft) : x)
                .Element(x => paddingRight != 0 ? x.PaddingRight(paddingRight) : x)
                .Element(x => paddingHorizontal != 0 ? x.PaddingHorizontal(paddingHorizontal) : x)
                .Element(x => paddingVertical != 0 ? x.PaddingVertical(paddingVertical) : x)
                .Element(x => background != "" ? x.Background(background) : x);`

**Comment by MarcinZiabek** [comment, marked as answer]
Thank you for the kind words, and apologies this went unanswered for so long!

Your general approach is correct — describing the design as data and applying it through extension methods is exactly the recommended pattern. A few refinements that remove most of the repetitive checks:

1) Zero-value calls are safe. Methods like `PaddingTop(0)` have no visual effect, so the conditional wrapping is unnecessary:

    ```csharp
    return parent
        .PaddingTop(design?.Padding?.Top ?? 0)
        .PaddingBottom(design?.Padding?.Bottom ?? 0)
        .PaddingLeft(design?.Padding?.Left ?? 0)
        .PaddingRight(design?.Padding?.Right ?? 0);
    ```    

2) Model alignment as an enum instead of boolean flags. This prevents invalid combinations (e.g. Center + Right both true) and reduces the checks to a single switch expression:

    ```csharp
    public static IContainer ApplyAlignment(this IContainer container, HorizontalAlignment? alignment) => alignment switch
    {
        HorizontalAlignment.Center => container.AlignCenter(),
        HorizontalAlignment.Right => container.AlignRight(),
        _ => container
    };
    ```

3) For truly optional styles, the `.Element(x => condition ? x.SomeStyle() : x)` pattern you used is the idiomatic solution — keeping it inside shared extension methods, as you did, is the right call.

---

Since this question was asked, the documentation gained a dedicated section covering these scenarios:

[Conditional formatting](https://www.questpdf.com/concepts/dynamic-composition/conditional-formatting.html) — applying styles based on values or settings,
[Configurable tables](https://www.questpdf.com/concepts/dynamic-composition/configurable-tables.html) — driving table structure and style from configuration objects (very close to your use case),
[Extension methods](https://www.questpdf.com/concepts/code-patterns/extension-methods.html) and [Content styling](https://www.questpdf.com/concepts/code-patterns/content-styling.html) — structuring reusable style helpers.

Closing as answered — feel free to follow up if anything is unclear.
